doc/develguide.rst

```diff
@@ -0,0 +1,656 @@
+
+Each Java type has a C++ wrapper class.  These classes provide a number of methods.
+Primitives each have their own unit type wrapper.  Object, arrays, and class 
+instances share a C++ wrapper type.  Special instances are used for 
+``java.lang.Object`` and ``java.lang.Class``.  The type wrapper are named for the class
+they wrap such as ``JPIntType``.  
+
+Type conversion
+++++++++++++++++
+
+For type conversion, a C++ class wrapper provides four methods.
+
+``canConvertToJava``
+  This method must consult the supplied Python object to determine the type
+  and then make a determination of whether a conversion is possible. 
+  It reports ``none_`` if there is no possible conversion, ``explicit_`` if the
+  conversion is only acceptable if forced such as returning from a proxy,
+  ``implicit_`` if the conversion is possible and acceptable as part of an 
+  method call, or ``exact_`` if this type converts without ambiguity.  It is excepted
+  to check for something that is already a Java resource of the correct type
+  such as ``JPValue``, or something this is implementing the behavior as an interface
+  in the form of a ``JPProxy``.
+
+``convertToJava``
+  This method consults the type and produces a conversion.  The order of the match
+  should be identical to the ``canConvertToJava``.  It should also handle values and
+  proxies.
+
+``convertToPythonObject``
+  This method takes a jvalue union and converts it to the corresponding 
+  Python wrapper instance.  
+
+``getValueFromObject``
+  This converts a Java object into a ``JPValue`` corresponding.  This unboxes
+  primitives.
+
+Array conversion
+++++++++++++++++++
+
+In addition to converting single objects, the type rewrappers also serve as the
+gateway to working with arrays of the specified type.  Five methods are used to 
+work with arrays:  ``newArrayInstance``, ``getArrayRange``, ``setArrayRange``, 
+``getArrayItem``, and ``setArrayItem``.  
+
+Invocation and Fields
+++++++++++++++++++++++
+
+To convert a return type produced from a Java call, each type needs to be
+able to invoke a method with that return type.  This corresponses the underlying
+JNI design.  The methods invoke and invokeStatic are used for this purpose.  
+Similarly accessing fields requires type conversion using the methods
+``getField`` and ``setField``.
+
+Instance verses Type wrappers
++++++++++++++++++++++++++++++++
+
+Instances of individual Java classes are made from ``JPClass``.  However, two
+special sets of conversion rules are required.  These are in the form 
+of specializations ``JPObjectBaseClass`` and ``JPClassBaseClass`` corresponding
+to ``java.lang.Object`` and ``java.lang.Class``.
+
+Support classes
+~~~~~~~~~~~~~~~
+
+In addition to the type wrappers, there are several support classes. These are:
+
+``JPTypeManager``
+  The typemanager serves as a dict for all type wrappers created during the 
+  operation.  
+
+``JPReferenceQueue``
+  Lifetime manager for Java and Python objects.
```

Comment:
Is the reference queue still required as you wrote that every object is wrapped in some kind of smart pointer?

Reply:
forget about this comment... I should have read further.